packages/blocks/src/api/raw-handling/html-formatting-remover.js

```diff
@@ -0,0 +1,87 @@
+/**
+ * Internal dependencies
+ */
+import { isPhrasingContent } from './phrasing-content';
+
+function getSibling( node, which ) {
+	const sibling = node[ `${ which }Sibling` ];
+
+	if ( sibling ) {
+		return sibling;
+	}
+
+	const { parentNode } = node;
+
+	if ( ! parentNode ) {
+		return;
+	}
+
+	if ( ! isPhrasingContent( parentNode ) ) {
+		return;
+	}
+
+	return getSibling( parentNode, which );
+}
+
+function isFormattingSpace( character ) {
+	return character === ' ' || character === '\t' || character === '\n';
+}
+
+/**
+ * Removes spacing that formats HTML.
+ *
+ * @param {Node} node The node to be processed.
+ * @return {void}
+ */
+export default function( node ) {
+	if ( node.nodeType !== node.TEXT_NODE ) {
+		return;
+	}
+
+	// Ignore pre content.
+	if ( node.parentElement.closest( 'pre' ) ) {
+		return;
+	}
+
+	// First, replace any sequence of HTML formatting space with a single space.
+	let newData = node.data.replace( /[ \n\t]+/g, ' ' );
+
+	// Remove the leading space if the text element is at the start of a block,
```

Comment:
Could we link to a relevant reference resource on which you're basing this logic?

I know in the past I've used this article:

https://medium.com/@patrickbrosset/when-does-white-space-matter-in-html-b90e8a7cdd33

...which references this specification:

https://www.w3.org/TR/css-text-3/#white-space-processing

Reply:
Out of curiosity, I ran the example in the article through the logic here as an added test case, and it didn't match the expected result.

```js
it.only( 'should match expected result of "When Does White Space Matter" article', () => {
	const input = '<h1>   Hello \n' +
		'				<span> World!</span>	  </h1>';

	expect( deepFilterHTML( input, [ filter ] ) ).toEqual( '<h1>Hello <span>World!</span></h1>' );
} );
```

```
  ● HTMLFormattingRemover › should match expected result of "When Does White Space Matter" article

    expect(received).toEqual(expected)

    Difference:

    - Expected
    + Received

    - <h1>Hello <span>World!</span></h1>
    + <h1>Hello<span> World!</span></h1>
```

Reply:
That's interesting. I'll have a look.

Reply:
Fixed.

Reply:
The fix created another problem. Fixing now.

Reply:
Could be good to have this as a test case, or at least a new one which verifies that whatever change was made cannot be regressed.

Reply:
There was actually a test case, but with the wrong expectation. :)

Reply:
See https://github.com/WordPress/gutenberg/pull/17470/commits/29d2f307ce36caaa6266d74a6303a09225613b28#diff-15a464d6f6d7c014b560b68da1c51c7bR91